app/workers/publishing_api_document_republishing_worker.rb

```diff
@@ -27,6 +27,22 @@ def perform(document_id, bulk_publishing = false)
     end
   end
 
+  def preview_refreshed_editions(document_id)
```

Comment:
I agree about this not being a great fit for the worker. Could always move this on to a `DocumentRepublisher` service or similar, which would be quite a Whitehall-ish thing to do.  I do like the "actions" concept, that's a nice way to organise the workflow.

Another option might be to move some of this onto the `Document` model. You could probably have `Document#has_republishable_editions?` and `Document#republishing_actions` methods which could be called from either the worker or the controller. Personally I'd probably go that route, I like rich Active Record models rather than passing models around to services, but either would probably be an improvement on this living in the worker.

You could probably go even further and the actions start becoming command objects in their own right or something... 🤔 might be overkill though 😂

Reply:
I like the sound of shifting this to `Document` model. Though given the actions are essentially the names of private methods in the worker, would it be odd for the model to know/be tied to those? I suppose we could consider the worker's responsibility simply to be defining an implementation for (and running) the actions as defined in the model

Reply:
Yeah I haven't thought it through too much. Maybe therefore the actions do need to be first class concepts instead of just private methods. Anyway, don't let my philosophical musings get in the way of you getting a republishing UI into the world!

Reply:
And a fixed version: https://github.com/alphagov/whitehall/pull/8993/commits/697343f6af092e7cd6fa64b7c3f609ae00765b59

Reply:
Looks pretty good, could we possibly do without the metaprogramming? If I'm correct that the metaprogramming is there to provide memoisation then I suspect we could do without it, thinking about it, because Active Record already [caches queries](https://guides.rubyonrails.org/caching_with_rails.html#sql-caching). So I don't think we're gaining much from it.

Reply:
It looks like that caching is [within the context of controller actions](https://guides.rubyonrails.org/caching_with_rails.html#sql-caching:~:text=it%27s%20important%20to%20note%20that%20query%20caches%20are%20created%20at%20the%20start%20of%20an%20action%20and%20destroyed%20at%20the%20end%20of%20that%20action%20and%20thus%20persist%20only%20for%20the%20duration%20of%20the%20action). I'd imagine that would work for the use case in this PR, but this worker is also used outside of controllers, and sometimes with `perform_async` or `perform_async_in_queue` - I'm not sure the caching would apply in those contexts?

Reply:
I think Sidekiq also applies caching: https://github.com/sidekiq/sidekiq/wiki/Problems-and-Troubleshooting#activerecord-query-cache

You could probably have a play around using `binding.pry` or `byebug` to see what it actually does, maybe.

Reply:
Definitely looks like caching is happening based on the Rails server output, so I'm happy to remove the fancy memoisation! I'll sort out working those changes into the commit history